error: unknown type name ‘WINDOW’

system: CentOS 6.5

cd samtools-1.3.1
./configure --enable-plugins --enable-libcurl --with-plugin-path=$PWD/htslib-1.3.1
make all plugins-htslib

...
gcc -g -O2 -I. -Ihtslib-1.3.1  -c -o bam_tview.o bam_tview.c
gcc -g -O2 -I. -Ihtslib-1.3.1  -c -o bam_tview_curses.o bam_tview_curses.c
bam_tview_curses.c:46:5: error: unknown type name ‘WINDOW’
bam_tview_curses.c: In function ‘curses_init_colors’:
bam_tview_curses.c:107:22: error: ‘COLOR_WHITE’ undeclared (first use in this function)
...

config.h
`/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in -- template for config.h.  If you use configure, this file
   provides #defines reflecting your configuration choices.  If you don't
   run configure, suitable conservative defaults will be used.

   This template file can be updated with autoheader, but do so carefully
   as autoheader adds #defines such as PACKAGE_* that we don't want.  */

/* Define to 1 if a SysV or X/Open compatible Curses library is present */
#define HAVE_CURSES 1

/* Define to 1 if library supports color (enhanced functions) */
/* #undef HAVE_CURSES_COLOR */

/* Define to 1 if library supports X/Open Enhanced functions */
/* #undef HAVE_CURSES_ENHANCED */

/* Define to 1 if <curses.h> is present */
/* #undef HAVE_CURSES_H */

/* Define to 1 if library supports certain obsolete features */
/* #undef HAVE_CURSES_OBSOLETE */

/* Define to 1 if the Ncurses library is present */
/* #undef HAVE_NCURSES */

/* Define to 1 if the NcursesW library is present */
#define HAVE_NCURSESW 1

/* Define to 1 if <ncursesw/curses.h> is present */
/* #undef HAVE_NCURSESW_CURSES_H */

/* Define to 1 if <ncursesw.h> is present */
/* #undef HAVE_NCURSESW_H */

/* Define to 1 if <ncurses/curses.h> is present */
/* #undef HAVE_NCURSES_CURSES_H */

/* Define to 1 if <ncurses.h> is present */
/* #undef HAVE_NCURSES_H */

/* Enable large inode numbers on Mac OS X 10.5.  */
#ifndef _DARWIN_USE_64_BIT_INODE
# define _DARWIN_USE_64_BIT_INODE 1
#endif

/* Number of bits in a file offset, on hosts where this is settable. */
/* #undef _FILE_OFFSET_BITS */

/* Define for large files, on AIX-style hosts. */
/* #undef _LARGE_FILES */
`

config.log
`This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by Samtools configure 1.3.1, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ./configure --enable-plugins --enable-libcurl --with-plugin-path=/home/zhuweiwei/bwadata/samtools-1.3.1/htslib-1.3.1

## --------- ##
## Platform. ##
## --------- ##

hostname = vsc-server
uname -m = x86_64
uname -r = 2.6.32-431.el6.x86_64
uname -s = Linux
uname -v = #1 SMP Fri Nov 22 03:15:09 UTC 2013

/usr/bin/uname -p = unknown
/bin/uname -X     = unknown

/bin/arch              = x86_64
/usr/bin/arch -k       = unknown
/usr/convex/getsysinfo = unknown
/usr/bin/hostinfo      = unknown
/bin/machine           = unknown
/usr/bin/oslevel       = unknown
/bin/universe          = unknown

PATH: /opt/centos/devtoolset-1.0/root/usr/bin/
PATH: /usr/lib64/qt-3.3/bin
PATH: /usr/local/bin
PATH: /bin
PATH: /usr/bin
PATH: /usr/local/sbin
PATH: /usr/sbin
PATH: /sbin
PATH: /home/zhuweiwei/bwadata/sratoolkit.2.5.7-ubuntu64/bin
PATH: /home/zhuweiwei/bin


## ----------- ##
## Core tests. ##
## ----------- ##

configure:1902: checking for gcc
configure:1918: found /opt/centos/devtoolset-1.0/root/usr/bin//gcc
configure:1929: result: gcc
configure:2158: checking for C compiler version
configure:2167: gcc --version >&5
gcc (GCC) 4.7.0 20120507 (Red Hat 4.7.0-5)
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

configure:2178: $? = 0
configure:2167: gcc -v >&5
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/opt/centos/devtoolset-1.0/root/usr/libexec/gcc/x86_64-redhat-linux/4.7.0/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/opt/centos/devtoolset-1.0/root/usr --mandir=/opt/centos/devtoolset-1.0/root/usr/share/man --infodir=/opt/centos/devtoolset-1.0/root/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --disable-build-with-cxx --disable-build-poststage1-with-cxx --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --enable-languages=c,c++,lto --enable-plugin --with-linker-hash-style=gnu --enable-initfini-array --disable-libgcj --with-ppl --with-cloog --with-mpc=/home/centos/rpm/BUILD/gcc-4.7.0-20120507/obj-x86_64-redhat-linux/mpc-install --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.7.0 20120507 (Red Hat 4.7.0-5) (GCC) 
configure:2178: $? = 0
configure:2167: gcc -V >&5
gcc: error: unrecognized command line option '-V'
gcc: fatal error: no input files
compilation terminated.
configure:2178: $? = 4
configure:2167: gcc -qversion >&5
gcc: error: unrecognized command line option '-qversion'
gcc: fatal error: no input files
compilation terminated.
configure:2178: $? = 4
configure:2198: checking whether the C compiler works
configure:2220: gcc    conftest.c  >&5
configure:2224: $? = 0
configure:2272: result: yes
configure:2275: checking for C compiler default output file name
configure:2277: result: a.out
configure:2283: checking for suffix of executables
configure:2290: gcc -o conftest    conftest.c  >&5
configure:2294: $? = 0
configure:2316: result: 
configure:2338: checking whether we are cross compiling
configure:2346: gcc -o conftest    conftest.c  >&5
configure:2350: $? = 0
configure:2357: ./conftest
configure:2361: $? = 0
configure:2376: result: no
configure:2381: checking for suffix of object files
configure:2403: gcc -c   conftest.c >&5
configure:2407: $? = 0
configure:2428: result: o
configure:2432: checking whether we are using the GNU C compiler
configure:2451: gcc -c   conftest.c >&5
configure:2451: $? = 0
configure:2460: result: yes
configure:2469: checking whether gcc accepts -g
configure:2489: gcc -c -g  conftest.c >&5
configure:2489: $? = 0
configure:2530: result: yes
configure:2547: checking for gcc option to accept ISO C89
configure:2610: gcc  -c -g -O2  conftest.c >&5
configure:2610: $? = 0
configure:2623: result: none needed
configure:2652: checking for special C compiler options needed for large files
configure:2697: result: no
configure:2703: checking for _FILE_OFFSET_BITS value needed for large files
configure:2728: gcc -c -g -O2  conftest.c >&5
configure:2728: $? = 0
configure:2760: result: no
configure:2888: checking location of HTSlib source tree
configure:2909: result: htslib-1.3.1
configure:3127: checking for NcursesW wide-character library
configure:3151: gcc -o conftest -g -O2   conftest.c  -lncursesw >&5
configure:3151: $? = 0
configure:3160: result: yes
configure:3180: checking for working ncursesw/curses.h
configure:3213: gcc -o conftest -g -O2   conftest.c  -lncursesw >&5
/opt/centos/devtoolset-1.0/root/usr/libexec/gcc/x86_64-redhat-linux/4.7.0/ld: /tmp/cceXsODk.o: undefined reference to symbol 'stdscr'
/opt/centos/devtoolset-1.0/root/usr/libexec/gcc/x86_64-redhat-linux/4.7.0/ld: note: 'stdscr' is defined in DSO /lib64/libtinfo.so.5 so try adding it to the linker command line
/lib64/libtinfo.so.5: could not read symbols: Invalid operation
collect2: error: ld returned 1 exit status
configure:3213: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "Samtools"
| #define PACKAGE_TARNAME "samtools"
| #define PACKAGE_VERSION "1.3.1"
| #define PACKAGE_STRING "Samtools 1.3.1"
| #define PACKAGE_BUGREPORT "samtools-help@lists.sourceforge.net"
| #define PACKAGE_URL "http://www.htslib.org/"
| #define HAVE_NCURSESW 1
| #define HAVE_CURSES 1
| /* end confdefs.h.  */
| 
|                         #define _XOPEN_SOURCE_EXTENDED 1
|                         #include <ncursesw/curses.h>
| 
| int
| main ()
| {
| 
|                         chtype a = A_BOLD;
|                         int b = KEY_LEFT;
|                         chtype c = COLOR_PAIR(1) & A_COLOR;
|                         attr_t d = WA_NORMAL;
|                         cchar_t e;
|                         wint_t f;
|                         int g = getattrs(stdscr);
|                         int h = getcurx(stdscr) + getmaxx(stdscr);
|                         initscr();
|                         init_pair(1, COLOR_WHITE, COLOR_RED);
|                         wattr_set(stdscr, d, 0, NULL);
|                         wget_wch(stdscr, &f);
| 
|   ;
|   return 0;
| }
configure:3222: result: no
configure:3244: checking for working ncursesw.h
configure:3277: gcc -o conftest -g -O2   conftest.c  -lncursesw >&5
conftest.c:13:46: fatal error: ncursesw.h: No such file or directory
compilation terminated.
configure:3277: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "Samtools"
| #define PACKAGE_TARNAME "samtools"
| #define PACKAGE_VERSION "1.3.1"
| #define PACKAGE_STRING "Samtools 1.3.1"
| #define PACKAGE_BUGREPORT "samtools-help@lists.sourceforge.net"
| #define PACKAGE_URL "http://www.htslib.org/"
| #define HAVE_NCURSESW 1
| #define HAVE_CURSES 1
| /* end confdefs.h.  */
| 
|                         #define _XOPEN_SOURCE_EXTENDED 1
|                         #include <ncursesw.h>
| 
| int
| main ()
| {
| 
|                         chtype a = A_BOLD;
|                         int b = KEY_LEFT;
|                         chtype c = COLOR_PAIR(1) & A_COLOR;
|                         attr_t d = WA_NORMAL;
|                         cchar_t e;
|                         wint_t f;
|                         int g = getattrs(stdscr);
|                         int h = getcurx(stdscr) + getmaxx(stdscr);
|                         initscr();
|                         init_pair(1, COLOR_WHITE, COLOR_RED);
|                         wattr_set(stdscr, d, 0, NULL);
|                         wget_wch(stdscr, &f);
| 
|   ;
|   return 0;
| }
configure:3286: result: no
configure:3308: checking for working ncurses.h
configure:3341: gcc -o conftest -g -O2   conftest.c  -lncursesw >&5
/opt/centos/devtoolset-1.0/root/usr/libexec/gcc/x86_64-redhat-linux/4.7.0/ld: /tmp/ccPm37nr.o: undefined reference to symbol 'stdscr'
/opt/centos/devtoolset-1.0/root/usr/libexec/gcc/x86_64-redhat-linux/4.7.0/ld: note: 'stdscr' is defined in DSO /lib64/libtinfo.so.5 so try adding it to the linker command line
/lib64/libtinfo.so.5: could not read symbols: Invalid operation
collect2: error: ld returned 1 exit status
configure:3341: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "Samtools"
| #define PACKAGE_TARNAME "samtools"
| #define PACKAGE_VERSION "1.3.1"
| #define PACKAGE_STRING "Samtools 1.3.1"
| #define PACKAGE_BUGREPORT "samtools-help@lists.sourceforge.net"
| #define PACKAGE_URL "http://www.htslib.org/"
| #define HAVE_NCURSESW 1
| #define HAVE_CURSES 1
| /* end confdefs.h.  */
| 
|                         #define _XOPEN_SOURCE_EXTENDED 1
|                         #include <ncurses.h>
| 
| int
| main ()
| {
| 
|                         chtype a = A_BOLD;
|                         int b = KEY_LEFT;
|                         chtype c = COLOR_PAIR(1) & A_COLOR;
|                         attr_t d = WA_NORMAL;
|                         cchar_t e;
|                         wint_t f;
|                         int g = getattrs(stdscr);
|                         int h = getcurx(stdscr) + getmaxx(stdscr);
|                         initscr();
|                         init_pair(1, COLOR_WHITE, COLOR_RED);
|                         wattr_set(stdscr, d, 0, NULL);
|                         wget_wch(stdscr, &f);
| 
|   ;
|   return 0;
| }
configure:3350: result: no
configure:3374: WARNING: could not find a working ncursesw/curses.h, ncursesw.h or ncurses.h
configure:3866: checking for zlib.h
configure:3866: gcc -c -g -O2  conftest.c >&5
configure:3866: $? = 0
configure:3866: result: yes
configure:3875: checking for inflate in -lz
configure:3900: gcc -o conftest -g -O2   conftest.c -lz   >&5
configure:3900: $? = 0
configure:3909: result: yes
configure:4051: creating ./config.status

## ---------------------- ##
## Running config.status. ##
## ---------------------- ##

This file was extended by Samtools config.status 1.3.1, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  CONFIG_FILES    = 
  CONFIG_HEADERS  = 
  CONFIG_LINKS    = 
  CONFIG_COMMANDS = 
  $ ./config.status 

on vsc-server

config.status:795: creating config.mk
config.status:795: creating config.h
configure:5276: === configuring in htslib-1.3.1 (/home/zhuweiwei/bwadata/samtools-1.3.1/htslib-1.3.1)
configure:5339: running /bin/sh ./configure --disable-option-checking '--prefix=/usr/local'  '--enable-plugins' '--enable-libcurl' '--with-plugin-path=/home/zhuweiwei/bwadata/samtools-1.3.1/htslib-1.3.1' --cache-file=/dev/null --srcdir=.

## ---------------- ##
## Cache variables. ##
## ---------------- ##

ac_cv_c_compiler_gnu=yes
ac_cv_env_CC_set=
ac_cv_env_CC_value=
ac_cv_env_CFLAGS_set=
ac_cv_env_CFLAGS_value=
ac_cv_env_CPPFLAGS_set=
ac_cv_env_CPPFLAGS_value=
ac_cv_env_CURSES_LIB_set=
ac_cv_env_CURSES_LIB_value=
ac_cv_env_LDFLAGS_set=
ac_cv_env_LDFLAGS_value=
ac_cv_env_LIBS_set=
ac_cv_env_LIBS_value=
ac_cv_env_build_alias_set=
ac_cv_env_build_alias_value=
ac_cv_env_host_alias_set=
ac_cv_env_host_alias_value=
ac_cv_env_target_alias_set=
ac_cv_env_target_alias_value=
ac_cv_header_zlib_h=yes
ac_cv_lib_z_inflate=yes
ac_cv_objext=o
ac_cv_prog_ac_ct_CC=gcc
ac_cv_prog_cc_c89=
ac_cv_prog_cc_g=yes
ac_cv_sys_file_offset_bits=no
ac_cv_sys_largefile_CC=no
ax_cv_curses=yes
ax_cv_curses_color=no
ax_cv_curses_enhanced=no
ax_cv_curses_obsolete=no
ax_cv_curses_which=ncursesw
ax_cv_header_ncurses_h_with_ncursesw=no
ax_cv_header_ncursesw_curses_h=no
ax_cv_header_ncursesw_h=no
ax_cv_htslib=yes
ax_cv_htslib_which=source
ax_cv_ncursesw=yes

## ----------------- ##
## Output variables. ##
## ----------------- ##

CC='gcc'
CFLAGS='-g -O2'
CPPFLAGS=''
CURSES_LIB='-lncursesw'
DEFS='-DHAVE_CONFIG_H'
ECHO_C=''
ECHO_N='-n'
ECHO_T=''
EXEEXT=''
HTSDIR='htslib-1.3.1'
HTSLIB_CPPFLAGS='-Ihtslib-1.3.1'
HTSLIB_LDFLAGS='-Lhtslib-1.3.1'
Hinstall='#'
Hsource=''
LDFLAGS=''
LIBOBJS=''
LIBS=''
LTLIBOBJS=''
OBJEXT='o'
PACKAGE_BUGREPORT='samtools-help@lists.sourceforge.net'
PACKAGE_NAME='Samtools'
PACKAGE_STRING='Samtools 1.3.1'
PACKAGE_TARNAME='samtools'
PACKAGE_URL='http://www.htslib.org/'
PACKAGE_VERSION='1.3.1'
PATH_SEPARATOR=':'
SHELL='/bin/sh'
ac_ct_CC='gcc'
bindir='${exec_prefix}/bin'
build_alias=''
datadir='${datarootdir}'
datarootdir='${prefix}/share'
docdir='${datarootdir}/doc/${PACKAGE_TARNAME}'
dvidir='${docdir}'
exec_prefix='${prefix}'
host_alias=''
htmldir='${docdir}'
includedir='${prefix}/include'
infodir='${datarootdir}/info'
libdir='${exec_prefix}/lib'
libexecdir='${exec_prefix}/libexec'
localedir='${datarootdir}/locale'
localstatedir='${prefix}/var'
mandir='${datarootdir}/man'
oldincludedir='/usr/include'
pdfdir='${docdir}'
prefix='/usr/local'
program_transform_name='s,x,x,'
psdir='${docdir}'
sbindir='${exec_prefix}/sbin'
sharedstatedir='${prefix}/com'
subdirs=' htslib-1.3.1'
sysconfdir='${prefix}/etc'
target_alias=''

## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "Samtools"
#define PACKAGE_TARNAME "samtools"
#define PACKAGE_VERSION "1.3.1"
#define PACKAGE_STRING "Samtools 1.3.1"
#define PACKAGE_BUGREPORT "samtools-help@lists.sourceforge.net"
#define PACKAGE_URL "http://www.htslib.org/"
#define HAVE_NCURSESW 1
#define HAVE_CURSES 1
#define HAVE_LIBZ 1

configure: exit 0
`

--- a/bam_tview_curses.c
+++ b/bam_tview_curses.c
@@ -267,7 +267,9 @@ static int curses_loop(tview_t* tv) {
             case 'K': tv->row_shift += 20; break;
             case KEY_BACKSPACE:
             case '\177': pos -= tv->mcol; break;
+#ifdef KEY_RESIZE
             case KEY_RESIZE: getmaxyx(stdscr, tv->mrow, tv->mcol); break;
+#endif
             default: continue;
         }
         if (pos < 0) pos = 0;

--- a/configure.ac
+++ b/configure.ac
@@ -30,7 +30,7 @@ AC_CONFIG_SRCDIR([bamtk.c])
 AC_CONFIG_HEADERS([config.h])
 
 m4_include([m4/hts_prog_cc_warnings.m4])
-m4_include([m4/ax_with_curses.m4])
+m4_include([m4/hts_with_curses.m4])
 m4_include([m4/ax_with_htslib.m4])
 
 dnl Copyright notice to be copied into the generated configure script
@@ -88,7 +88,7 @@ AC_ARG_WITH([curses],
                   [omit curses support, so no curses library needed])])
 
 if test "$with_curses" != no; then
-  AX_WITH_CURSES
+  HTS_WITH_CURSES
   if test "$ax_cv_curses" != yes; then
     AC_MSG_ERROR([curses development files not found
 

--- a/m4/hts_with_curses.m4
+++ b/m4/hts_with_curses.m4
@@ -1,10 +1,11 @@
 # ===========================================================================
+#   Based on serial 15 of AX_WITH_CURSES from autoconf-archive
 #      http://www.gnu.org/software/autoconf-archive/ax_with_curses.html
 # ===========================================================================
 #
 # SYNOPSIS
 #
-#   AX_WITH_CURSES
+#   HTS_WITH_CURSES
 #
 # DESCRIPTION
 #
@@ -94,7 +95,7 @@
 #   If any Curses library will do (but one must be present and must support
 #   color), you could use:
 #
-#     AX_WITH_CURSES
+#     HTS_WITH_CURSES
 #     if test "x$ax_cv_curses" != xyes || test "x$ax_cv_curses_color" != xyes; then
 #         AC_MSG_ERROR([requires an X/Open-compatible Curses library with color])
 #     fi
@@ -149,12 +150,16 @@
 #   of AX_WITH_CURSES, as well as in the additional macros
 #   AX_WITH_CURSES_PANEL, AX_WITH_CURSES_MENU and AX_WITH_CURSES_FORM.
 #
+#   Based on serial 15 of AX_WITH_CURSES from autoconf-archive
+#      http://www.gnu.org/software/autoconf-archive/ax_with_curses.html
+#
 # LICENSE
 #
 #   Copyright (c) 2009 Mark Pulford <mark@kyne.com.au>
 #   Copyright (c) 2009 Damian Pietras <daper@daper.net>
 #   Copyright (c) 2012 Reuben Thomas <rrt@sc3d.org>
 #   Copyright (c) 2011 John Zaitseff <J.Zaitseff@zap.org.au>
+#   Copyright (c) 2020 Genome Research Limited
 #
 #   This program is free software: you can redistribute it and/or modify it
 #   under the terms of the GNU General Public License as published by the
@@ -182,10 +187,8 @@
 #   modified version of the Autoconf Macro, you may extend this special
 #   exception to the GPL to apply to your modified version as well.
 
-#serial 15
-
-AU_ALIAS([MP_WITH_CURSES], [AX_WITH_CURSES])
-AC_DEFUN([AX_WITH_CURSES], [
+AU_ALIAS([MP_WITH_CURSES], [HTS_WITH_CURSES])
+AC_DEFUN([HTS_WITH_CURSES], [
     AC_ARG_VAR([CURSES_LIB], [linker library for Curses, e.g. -lcurses])
     AC_ARG_WITH([ncurses], [AS_HELP_STRING([--with-ncurses],
         [force the use of Ncurses or NcursesW])],

--- a/m4/hts_with_curses.m4
+++ b/m4/hts_with_curses.m4
@@ -297,6 +297,14 @@ AC_DEFUN([HTS_WITH_CURSES], [
                 AC_DEFINE([HAVE_NCURSES_H],       [1], [Define to 1 if <ncurses.h> is present])
                 ax_cv_curses_which=ncursesw
             ])
+            HTS_NCURSESW_HEADERS_CHECK([for working ncurses/curses.h],
+                                       [ax_cv_header_ncurses_curses_h_with_ncursesw],
+                                       [ncurses/curses.h],
+                                       [test "x$ax_cv_curses_which" = xno])
+            AS_IF([test "x$ax_cv_header_ncurses_curses_h_with_ncursesw" = xyes], [
+                AC_DEFINE([HAVE_NCURSES_CURSES_H], [1], [Define to 1 if <ncurses/curses.h> is present])
+                ax_cv_curses_which=ncursesw
+            ])
             AS_IF([test "x$ax_cv_curses_which" = xncursesw], [
                 ax_cv_curses=yes
                 ax_cv_curses_enhanced=yes

--- a/m4/hts_with_curses.m4
+++ b/m4/hts_with_curses.m4
@@ -187,6 +187,62 @@
 #   modified version of the Autoconf Macro, you may extend this special
 #   exception to the GPL to apply to your modified version as well.
 
+dnl Check for ncursesw headers.  Expanded a lot due to the number of
+dnl places they may be stored.
+dnl
+dnl $1 is the message for AC_CACHE_CHECK
+dnl $2 is the output variable
+dnl $3 is the header file name
+dnl $4 is a test to see if the check is required
+m4_define([HTS_NCURSESW_HEADERS_CHECK],[
+    AS_IF([$4], [
+        AC_CACHE_CHECK([$1], [$2], [
+            AC_LINK_IFELSE([AC_LANG_PROGRAM([[
+                @%:@define _XOPEN_SOURCE_EXTENDED 1
+                @%:@include <$3>
+            ]], [[
+                chtype a = A_BOLD;
+                int b = KEY_LEFT;
+                chtype c = COLOR_PAIR(1) & A_COLOR;
+                attr_t d = WA_NORMAL;
+                cchar_t e;
+                wint_t f;
+                int g = getattrs(stdscr);
+                int h = getcurx(stdscr) + getmaxx(stdscr);
+                initscr();
+                init_pair(1, COLOR_WHITE, COLOR_RED);
+                wattr_set(stdscr, d, 0, NULL);
+                wget_wch(stdscr, &f);
+            ]])],
+            [$2=yes],
+            [$2=no])
+        ])
+    ])
+])
+
+dnl Check for ncurses headers
+dnl Parameters as for HTS_NCURSESW_HEADERS_CHECK
+
+m4_define([HTS_NCURSES_HEADERS_CHECK],[
+    AS_IF([$4], [
+        AC_CACHE_CHECK([$1], [$2], [
+            AC_LINK_IFELSE([AC_LANG_PROGRAM([[
+                @%:@include <$3>
+            ]], [[
+                chtype a = A_BOLD;
+                int b = KEY_LEFT;
+                chtype c = COLOR_PAIR(1) & A_COLOR;
+                int g = getattrs(stdscr);
+                int h = getcurx(stdscr) + getmaxx(stdscr);
+                initscr();
+                init_pair(1, COLOR_WHITE, COLOR_RED);
+            ]])],
+            [$2=yes],
+            [$2=no])
+        ])
+    ])
+])
+
 AU_ALIAS([MP_WITH_CURSES], [HTS_WITH_CURSES])
 AC_DEFUN([HTS_WITH_CURSES], [
     AC_ARG_VAR([CURSES_LIB], [linker library for Curses, e.g. -lcurses])
@@ -217,107 +273,47 @@ AC_DEFUN([HTS_WITH_CURSES], [
         ])
 
         AS_IF([test "x$ax_cv_ncursesw" = xyes], [
-            ax_cv_curses=yes
-            ax_cv_curses_which=ncursesw
-            CURSES_LIB="-lncursesw"
-            AC_DEFINE([HAVE_NCURSESW], [1], [Define to 1 if the NcursesW library is present])
-            AC_DEFINE([HAVE_CURSES],   [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
-
-            AC_CACHE_CHECK([for working ncursesw/curses.h], [ax_cv_header_ncursesw_curses_h], [
-                AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-                        @%:@define _XOPEN_SOURCE_EXTENDED 1
-                        @%:@include <ncursesw/curses.h>
-                    ]], [[
-                        chtype a = A_BOLD;
-                        int b = KEY_LEFT;
-                        chtype c = COLOR_PAIR(1) & A_COLOR;
-                        attr_t d = WA_NORMAL;
-                        cchar_t e;
-                        wint_t f;
-                        int g = getattrs(stdscr);
-                        int h = getcurx(stdscr) + getmaxx(stdscr);
-                        initscr();
-                        init_pair(1, COLOR_WHITE, COLOR_RED);
-                        wattr_set(stdscr, d, 0, NULL);
-                        wget_wch(stdscr, &f);
-                    ]])],
-                    [ax_cv_header_ncursesw_curses_h=yes],
-                    [ax_cv_header_ncursesw_curses_h=no])
-            ])
+            HTS_NCURSESW_HEADERS_CHECK([for working ncursesw/curses.h],
+                                       [ax_cv_header_ncursesw_curses_h],
+                                       [ncursesw/curses.h],
+                                       [test "x$ax_cv_curses_which" = xno])
             AS_IF([test "x$ax_cv_header_ncursesw_curses_h" = xyes], [
-                ax_cv_curses_enhanced=yes
-                ax_cv_curses_color=yes
-                ax_cv_curses_obsolete=yes
-                AC_DEFINE([HAVE_CURSES_ENHANCED],   [1], [Define to 1 if library supports X/Open Enhanced functions])
-                AC_DEFINE([HAVE_CURSES_COLOR],      [1], [Define to 1 if library supports color (enhanced functions)])
-                AC_DEFINE([HAVE_CURSES_OBSOLETE],   [1], [Define to 1 if library supports certain obsolete features])
                 AC_DEFINE([HAVE_NCURSESW_CURSES_H], [1], [Define to 1 if <ncursesw/curses.h> is present])
+                ax_cv_curses_which=ncursesw
             ])
-
-            AC_CACHE_CHECK([for working ncursesw.h], [ax_cv_header_ncursesw_h], [
-                AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-                        @%:@define _XOPEN_SOURCE_EXTENDED 1
-                        @%:@include <ncursesw.h>
-                    ]], [[
-                        chtype a = A_BOLD;
-                        int b = KEY_LEFT;
-                        chtype c = COLOR_PAIR(1) & A_COLOR;
-                        attr_t d = WA_NORMAL;
-                        cchar_t e;
-                        wint_t f;
-                        int g = getattrs(stdscr);
-                        int h = getcurx(stdscr) + getmaxx(stdscr);
-                        initscr();
-                        init_pair(1, COLOR_WHITE, COLOR_RED);
-                        wattr_set(stdscr, d, 0, NULL);
-                        wget_wch(stdscr, &f);
-                    ]])],
-                    [ax_cv_header_ncursesw_h=yes],
-                    [ax_cv_header_ncursesw_h=no])
-            ])
+            HTS_NCURSESW_HEADERS_CHECK([for working ncursesw.h],
+                                       [ax_cv_header_ncursesw_h],
+                                       [ncursesw.h],
+                                       [test "x$ax_cv_curses_which" = xno])
             AS_IF([test "x$ax_cv_header_ncursesw_h" = xyes], [
-                ax_cv_curses_enhanced=yes
-                ax_cv_curses_color=yes
-                ax_cv_curses_obsolete=yes
-                AC_DEFINE([HAVE_CURSES_ENHANCED], [1], [Define to 1 if library supports X/Open Enhanced functions])
-                AC_DEFINE([HAVE_CURSES_COLOR],    [1], [Define to 1 if library supports color (enhanced functions)])
-                AC_DEFINE([HAVE_CURSES_OBSOLETE], [1], [Define to 1 if library supports certain obsolete features])
                 AC_DEFINE([HAVE_NCURSESW_H],      [1], [Define to 1 if <ncursesw.h> is present])
+                ax_cv_curses_which=ncursesw
             ])
-
-            AC_CACHE_CHECK([for working ncurses.h], [ax_cv_header_ncurses_h_with_ncursesw], [
-                AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-                        @%:@define _XOPEN_SOURCE_EXTENDED 1
-                        @%:@include <ncurses.h>
-                    ]], [[
-                        chtype a = A_BOLD;
-                        int b = KEY_LEFT;
-                        chtype c = COLOR_PAIR(1) & A_COLOR;
-                        attr_t d = WA_NORMAL;
-                        cchar_t e;
-                        wint_t f;
-                        int g = getattrs(stdscr);
-                        int h = getcurx(stdscr) + getmaxx(stdscr);
-                        initscr();
-                        init_pair(1, COLOR_WHITE, COLOR_RED);
-                        wattr_set(stdscr, d, 0, NULL);
-                        wget_wch(stdscr, &f);
-                    ]])],
-                    [ax_cv_header_ncurses_h_with_ncursesw=yes],
-                    [ax_cv_header_ncurses_h_with_ncursesw=no])
-            ])
+            HTS_NCURSESW_HEADERS_CHECK([for working ncurses.h],
+                                       [ax_cv_header_ncurses_h_with_ncursesw],
+                                       [ncurses.h],
+                                       [test "x$ax_cv_curses_which" = xno])
             AS_IF([test "x$ax_cv_header_ncurses_h_with_ncursesw" = xyes], [
+                AC_DEFINE([HAVE_NCURSES_H],       [1], [Define to 1 if <ncurses.h> is present])
+                ax_cv_curses_which=ncursesw
+            ])
+            AS_IF([test "x$ax_cv_curses_which" = xncursesw], [
+                ax_cv_curses=yes
                 ax_cv_curses_enhanced=yes
                 ax_cv_curses_color=yes
                 ax_cv_curses_obsolete=yes
+                CURSES_LIB="-lncursesw"
+                AC_DEFINE([HAVE_NCURSESW], [1], [Define to 1 if the NcursesW library is present])
+                AC_DEFINE([HAVE_CURSES],   [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
                 AC_DEFINE([HAVE_CURSES_ENHANCED], [1], [Define to 1 if library supports X/Open Enhanced functions])
                 AC_DEFINE([HAVE_CURSES_COLOR],    [1], [Define to 1 if library supports color (enhanced functions)])
                 AC_DEFINE([HAVE_CURSES_OBSOLETE], [1], [Define to 1 if library supports certain obsolete features])
-                AC_DEFINE([HAVE_NCURSES_H],       [1], [Define to 1 if <ncurses.h> is present])
-            ])
-
-            AS_IF([test "x$ax_cv_header_ncursesw_curses_h" = xno && test "x$ax_cv_header_ncursesw_h" = xno && test "x$ax_cv_header_ncurses_h_with_ncursesw" = xno], [
-                AC_MSG_WARN([could not find a working ncursesw/curses.h, ncursesw.h or ncurses.h])
+              ], [
+                AS_IF([test "x$with_ncursesw" = xyes], [
+                    AC_MSG_ERROR([--with-ncursesw specified but could not find NcursesW header])
+                ], [
+                    AC_MSG_WARN([could not find a working ncursesw/curses.h, ncursesw.h or ncurses.h])
+                ])
             ])
         ])
     ])
@@ -336,60 +332,39 @@ AC_DEFUN([HTS_WITH_CURSES], [
         ])
 
         AS_IF([test "x$ax_cv_ncurses" = xyes], [
-            ax_cv_curses=yes
-            ax_cv_curses_which=ncurses
-            CURSES_LIB="-lncurses"
-            AC_DEFINE([HAVE_NCURSES], [1], [Define to 1 if the Ncurses library is present])
-            AC_DEFINE([HAVE_CURSES],  [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
-
-            AC_CACHE_CHECK([for working ncurses/curses.h], [ax_cv_header_ncurses_curses_h], [
-                AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-                        @%:@include <ncurses/curses.h>
-                    ]], [[
-                        chtype a = A_BOLD;
-                        int b = KEY_LEFT;
-                        chtype c = COLOR_PAIR(1) & A_COLOR;
-                        int g = getattrs(stdscr);
-                        int h = getcurx(stdscr) + getmaxx(stdscr);
-                        initscr();
-                        init_pair(1, COLOR_WHITE, COLOR_RED);
-                    ]])],
-                    [ax_cv_header_ncurses_curses_h=yes],
-                    [ax_cv_header_ncurses_curses_h=no])
-            ])
+            HTS_NCURSES_HEADERS_CHECK([for working ncurses/curses.h],
+                                      [ax_cv_header_ncurses_curses_h],
+                                      [ncurses/curses.h],
+                                      [test "x$ax_cv_curses_which" = xno])
             AS_IF([test "x$ax_cv_header_ncurses_curses_h" = xyes], [
-                ax_cv_curses_color=yes
-                ax_cv_curses_obsolete=yes
-                AC_DEFINE([HAVE_CURSES_COLOR],     [1], [Define to 1 if library supports color (enhanced functions)])
-                AC_DEFINE([HAVE_CURSES_OBSOLETE],  [1], [Define to 1 if library supports certain obsolete features])
                 AC_DEFINE([HAVE_NCURSES_CURSES_H], [1], [Define to 1 if <ncurses/curses.h> is present])
+                ax_cv_curses_which=ncurses
             ])
 
-            AC_CACHE_CHECK([for working ncurses.h], [ax_cv_header_ncurses_h], [
-                AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-                        @%:@include <ncurses.h>
-                    ]], [[
-                        chtype a = A_BOLD;
-                        int b = KEY_LEFT;
-                        chtype c = COLOR_PAIR(1) & A_COLOR;
-                        int g = getattrs(stdscr);
-                        int h = getcurx(stdscr) + getmaxx(stdscr);
-                        initscr();
-                        init_pair(1, COLOR_WHITE, COLOR_RED);
-                    ]])],
-                    [ax_cv_header_ncurses_h=yes],
-                    [ax_cv_header_ncurses_h=no])
-            ])
+            HTS_NCURSES_HEADERS_CHECK([for working ncurses.h],
+                                      [ax_cv_header_ncurses_h],
+                                      [ncurses.h],
+                                      [test "x$ax_cv_curses_which" = xno])
             AS_IF([test "x$ax_cv_header_ncurses_h" = xyes], [
+                AC_DEFINE([HAVE_NCURSES_H],       [1], [Define to 1 if <ncurses.h> is present])
+                ax_cv_curses_which=ncurses
+            ])
+
+            AS_IF([test "x$ax_cv_curses_which" = xncurses], [
+                ax_cv_curses=yes
                 ax_cv_curses_color=yes
                 ax_cv_curses_obsolete=yes
+                CURSES_LIB="-lncurses"
+                AC_DEFINE([HAVE_NCURSES], [1], [Define to 1 if the Ncurses library is present])
+                AC_DEFINE([HAVE_CURSES],  [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
                 AC_DEFINE([HAVE_CURSES_COLOR],    [1], [Define to 1 if library supports color (enhanced functions)])
                 AC_DEFINE([HAVE_CURSES_OBSOLETE], [1], [Define to 1 if library supports certain obsolete features])
-                AC_DEFINE([HAVE_NCURSES_H],       [1], [Define to 1 if <ncurses.h> is present])
-            ])
-
-            AS_IF([test "x$ax_cv_header_ncurses_curses_h" = xno && test "x$ax_cv_header_ncurses_h" = xno], [
-                AC_MSG_WARN([could not find a working ncurses/curses.h or ncurses.h])
+              ], [
+                AS_IF([test "x$with_ncurses" = xyes], [
+                    AC_MSG_ERROR([--with-ncurses specified but could not find Ncurses header])
+                ], [
+                    AC_MSG_WARN([could not find a working ncurses/curses.h or ncurses.h])
+                ])
             ])
         ])
     ])
@@ -409,12 +384,6 @@ AC_DEFUN([HTS_WITH_CURSES], [
         ])
 
         AS_IF([test "x$ax_cv_plaincurses" = xyes], [
-            ax_cv_curses=yes
-            ax_cv_curses_which=plaincurses
-            AS_IF([test "x$CURSES_LIB" = x], [
-                CURSES_LIB="-lcurses"
-            ])
-            AC_DEFINE([HAVE_CURSES], [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
 
             # Check for base conformance (and header file)
 
@@ -430,6 +399,12 @@ AC_DEFUN([HTS_WITH_CURSES], [
                     [ax_cv_header_curses_h=no])
             ])
             AS_IF([test "x$ax_cv_header_curses_h" = xyes], [
+                ax_cv_curses=yes
+                ax_cv_curses_which=plaincurses
+                AS_IF([test "x$CURSES_LIB" = x], [
+                    CURSES_LIB="-lcurses"
+                ])
+                AC_DEFINE([HAVE_CURSES], [1], [Define to 1 if a SysV or X/Open compatible Curses library is present])
                 AC_DEFINE([HAVE_CURSES_H], [1], [Define to 1 if <curses.h> is present])
 
                 # Check for X/Open Enhanced conformance

--- a/m4/hts_with_curses.m4
+++ b/m4/hts_with_curses.m4
@@ -495,6 +495,15 @@ AC_DEFUN([HTS_WITH_CURSES], [
         ])
     ])
 
+    AS_IF([test "x$ax_cv_curses_which" != xno], [
+        # Normally linking curses gets tinfo as well, but this doesn't
+        # always happen.  Check to see if it needs to be added explicitly.
+        AC_SEARCH_LIBS(cbreak, tinfo)
+        AS_CASE([$ac_cv_search_cbreak],
+                [-l*],[CURSES_LIB="$CURSES_LIB $ac_cv_search_cbreak"
+                       LIBS="$LIBS $ac_cv_search_cbreak"])
+    ])
+
     AS_IF([test "x$ax_cv_curses"          != xyes], [ax_cv_curses=no])
     AS_IF([test "x$ax_cv_curses_enhanced" != xyes], [ax_cv_curses_enhanced=no])
     AS_IF([test "x$ax_cv_curses_color"    != xyes], [ax_cv_curses_color=no])

